src/care/host_device_ptr.h

```diff
@@ -202,11 +202,9 @@ namespace care {
       /// Return the value at the given index
       ///
       template<typename Idx>
-      inline CARE_HOST_DEVICE T& operator[](const Idx i)
-#ifndef CARE_LEGACY_COMPATIBILITY_MODE
-      const
-#endif
-      { return MA::operator[](i); }
+      inline CARE_HOST_DEVICE T& operator[](const Idx i) const { 
```

Comment:
I believe this change should no longer be necessary if CARE_LEGACY_COMPATIBILITY_MODE is correctly defined now.

Reply:
It seems that CARE_LEGACY_COMPATIBILITY_MODE is designed to change other behavior, including loop fusion, "permuted connectivity", how a bunch of the care loops operate )see care.h and DefaulyMacros.h). Are you really ok with changing all of that behavior on all builds? 

I need this change to get the implicit conversion stuff working, which would apply to all builds. I was hesitant to change a bunch of other behaviors

Reply:
CARE_LEGACY_COMPATIBILITY_MODE is already (or we thought it was already) turned on for the normal icc toss release build. It should be turned off for all other builds. When it is on, the loop macros degrade to raw for loops, meaning there is no lambda capture, and the compiler can do a lot more optimization making things much faster. Also, we don't have to violate the normal const rules, which state that if you are returning a non-const reference the method should be non-const. Hence why this method is only const if we are not in legacy mode.

Reply:
I would undo those changes and try adding an overload:

#if defined(CARE_LEGACY_COMPATIBILITY_MODE)
inline CARE_HOST_DEVICE const T& operator[](const Idx i) const {
          return MA::operator[](i); 
}
#endif

Reply:
Undid change, added your bit of code. We will talk Monday

Reply:
You can take another look. I will keep LEGACY mode disabled in ale3d for now; that will be a separate pull request